### PR TITLE
Feature/login subscribe runRules

### PR DESCRIPTION
**Pull request by ofiralexi**
Description: pull request matching issues #49 and #50 . Added intent handlers that access the DB.

### CODE REVIEW

**Comment by ShaiZeevi** [comment]
The results from runRules:
image

**Comment by ShaiZeevi** [comment]
Also @ofiralexi I've changed login to include a password as we discussed on the phone (@yoavz1997 approved).
If you want to change it in the next couple of days I'm open to it

File: src/main/java/FuncTest/TestFunc/Function.java [hunk 1]
```diff
@@ -32,21 +35,64 @@ public HttpResponseMessage run(
...
 		return utils.createWebhookResponseContent("what is this intent?", s);
 	}
+	
+	private HttpResponseMessage applyRules(JSONObject queryResult, HttpRequestMessage<Optional<String>> s,
+			ExecutionContext c) {
+		// TODO Auto-generated method stub
+		String uname = utils.getStringUserParamFromContext(queryResult, "username");
+		String passwd = utils.getStringUserParamFromContext(queryResult, "password");
+		
+		c.getLogger().info("================ username = " + uname + "================");
```
Comment (yoavz1997) [inline src/main/java/FuncTest/TestFunc/Function.java]: Please remove before final merge of the branch

File: src/main/java/FuncTest/TestFunc/Function.java [hunk 2]
```diff
@@ -32,21 +35,64 @@ public HttpResponseMessage run(
...
+		List<String> requiredParameterNames = Arrays.asList("username", "password");
+		
+		if (!utils.allParametersArePresent(parameters, requiredParameterNames)) 
+			return utils.createWebhookResponseContent("Please enter user credentials to subscribe with.", s);
+		
+		return utils.createWebhookResponseContent(
+				(new subscribeHandler(parameters.getString("username"),parameters.getString("password"))).subscribe(c), s);
+	}
+	
+	private HttpResponseMessage checkLoginName(JSONObject queryResult, HttpRequestMessage<Optional<String>> s,
```
Comment (yoavz1997) [inline src/main/java/FuncTest/TestFunc/Function.java]: Duplicate of the previous function, so it seems to be redundant.

File: src/main/java/rule/loginHandler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	
+	public loginHandler(String username, String password) {
+		this.username = username;
+		this.password = password;
+	}
+	
+	public String checkUserNameExists() {
+		Connection connection = null;
+		StringBuilder jsonResult = new StringBuilder();
+		try {
```
Comment (yoavz1997) [inline src/main/java/rule/loginHandler.java]: Please use try with resource https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: src/main/java/FuncTest/TestFunc/Function.java (hunk 1 above)
Comment (ShaiZeevi) [reply src/main/java/FuncTest/TestFunc/Function.java]: I've removed the logging

File: src/main/java/FuncTest/TestFunc/Function.java (hunk 2 above)
Comment (ShaiZeevi) [reply src/main/java/FuncTest/TestFunc/Function.java]: I disagree with that. The former checks the user has an entered details and then performs a "subscription" (writes to the DB). The latter, check the user has entered details and then checks they are correct (reads from the DB).
Both operations are done through intermediary classes.

I believe though they are similar each possesses a distinct "source of truth" (as we learned in Software Design) and so it's fine in my opinion

File: src/main/java/rule/loginHandler.java (hunk 3 above)
Comment (ShaiZeevi) [reply src/main/java/rule/loginHandler.java]: Done

File: src/main/java/FuncTest/TestFunc/Function.java (hunk 1 above)
Comment (ShaiZeevi) [reply src/main/java/FuncTest/TestFunc/Function.java]: Done

File: src/main/java/FuncTest/TestFunc/Function.java (hunk 2 above)
Comment (ShaiZeevi) [reply src/main/java/FuncTest/TestFunc/Function.java]: I do not agree as explained in the conversation under the "files changed" tab

File: src/main/java/rule/loginHandler.java (hunk 3 above)
Comment (ShaiZeevi) [reply src/main/java/rule/loginHandler.java]: Done

**Pull request merged**